plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformer.kt

```diff
@@ -0,0 +1,595 @@
+
+    private fun IrCall.inlineAtomicFunction(atomicType: IrType, accessors: List<IrExpression>): IrCall {
+        val valueArguments = getValueArguments()
+        val functionName = getAtomicFunctionName()
+        val runtimeFunction = getRuntimeFunctionSymbol(functionName, atomicType)
+        return buildCall(
+            target = runtimeFunction,
+            type = type,
+            origin = IrStatementOrigin.INVOKE,
+            typeArguments = if (runtimeFunction.owner.typeParameters.size == 1) listOf(atomicType) else emptyList(),
+            valueArguments = valueArguments + accessors
+        )
+    }
+
+    private fun IrStatement.transformStatement(parentDeclaration: IrDeclaration, lambda: IrFunction? = null) =
+        when (this) {
+            is IrExpression -> transformAtomicFunctionCall(parentDeclaration, lambda)
+            is IrVariable -> {
+                apply { initializer = initializer?.transformAtomicFunctionCall(parentDeclaration, lambda) }
+            }
+            else -> this
+        }
+
+    private fun IrExpression.transformAtomicFunctionCall(parentDeclaration: IrDeclaration, lambda: IrFunction? = null): IrExpression {
+        // erase unchecked cast to the Atomic* type
+        if (this is IrTypeOperatorCall && operator == IrTypeOperator.CAST && typeOperand.isAtomicValueType()) {
+            return argument
+        }
+        if (isAtomicValueInitializer()) {
+            return eraseAtomicConstructor(parentDeclaration, lambda)
+        }
+        when (this) {
+            is IrTypeOperatorCall -> {
+                return apply { argument = argument.transformAtomicFunctionCall(parentDeclaration, lambda) }
+            }
+            is IrStringConcatenationImpl -> {
+                return apply {
+                    arguments.forEachIndexed { i, arg ->
+                        arguments[i] = arg.transformAtomicFunctionCall(parentDeclaration, lambda)
+                    }
+                }
+            }
+            is IrReturn -> {
+                if (parentDeclaration.isTransformedAtomicExtensionFunction() && (parentDeclaration as IrFunction).isInline &&
+                    returnTargetSymbol !== parentDeclaration.symbol) {
+                    return IrReturnImpl(
+                        startOffset,
+                        endOffset,
+                        type,
+                        parentDeclaration.symbol,
+                        value.transformAtomicFunctionCall(parentDeclaration, lambda)
+                    )
+                }
+                return apply { value = value.transformAtomicFunctionCall(parentDeclaration, lambda) }
+            }
+            is IrSetValue -> {
+                return apply { value = value.transformAtomicFunctionCall(parentDeclaration, lambda) }
+            }
+            is IrSetField -> {
+                return apply { value = value.transformAtomicFunctionCall(parentDeclaration, lambda) }
+            }
+            is IrIfThenElseImpl -> {
```

Comment:
I would avoid using `Impl` here. It's part of the internal implementation of the compiler.